lib/spack/spack/hooks/permissions_setters.py

```diff
@@ -4,54 +4,18 @@
 # SPDX-License-Identifier: (Apache-2.0 OR MIT)
 
 import os
-import stat
-
-from llnl.util.filesystem import chmod_x, chgrp
-
-from spack.package_prefs import get_package_permissions, get_package_group
-from spack.package_prefs import get_package_dir_permissions
-from spack.error import SpackError
-
-
-def forall_files(path, fn, args, dir_args=None):
-    """Apply function to all files in directory, with file as first arg.
-
-    Does not apply to the root dir. Does not apply to links"""
-    # os.walk explicitly set not to follow links
-    for root, dirs, files in os.walk(path, followlinks=False):
-        for d in dirs:
-            if not os.path.islink(os.path.join(root, d)):
-                if dir_args:
-                    fn(os.path.join(root, d), *dir_args)
-                else:
-                    fn(os.path.join(root, d), *args)
-        for f in files:
-            if not os.path.islink(os.path.join(root, f)):
-                fn(os.path.join(root, f), *args)
-
-
-def chmod_real_entries(path, perms):
-    # Don't follow links so we don't change things outside the prefix
-    if not os.path.islink(path):
-        mode = os.stat(path).st_mode
-        perms |= mode & (stat.S_ISUID | stat.S_ISGID | stat.S_ISVTX)
-        if perms & stat.S_ISUID and perms & stat.S_IWGRP:
-            raise InvalidPermissionsError(
-                'Attempting to set suid with world writable')
-        chmod_x(path, perms)
+import spack.util.file_permissions as fp
 
 
 def post_install(spec):
     if not spec.external:
-        perms = get_package_permissions(spec)
-        dir_perms = get_package_dir_permissions(spec)
-        group = get_package_group(spec)
-
-        forall_files(spec.prefix, chmod_real_entries, [perms], [dir_perms])
-
-        if group:
-            forall_files(spec.prefix, chgrp, [group])
-
-
-class InvalidPermissionsError(SpackError):
-    """Error class for invalid permission setters"""
+        fp.set_permissions_by_spec(spec.prefix, spec)
+
+        # os.walk explicitly set not to follow links
+        for root, dirs, files in os.walk(spec.prefix, followlinks=False):
```

Comment:
FYI. os.walk is notorious for being slow, especially in a "large" directory tree.

Reply:
Newer versions of Spack use scandir, we can backport it if we run into performance issues

Reply:
We might consider vendoring scandir, as there are a number of places we could improve with this. Binary package creation, permission setting, etc. I implemented something like this for checking for package updates and regenerating the virtual and patch caches:

https://github.com/spack/spack/blob/develop/lib/spack/spack/repo.py#L115

Maybe that is replaceable too. Spack currently does too many filesystem ops and it’d be cool to reduce startup overhead with this if it makes sense.